philander/stc311.py

```diff
@@ -0,0 +1,665 @@
+        else:
+            ret = BatLevel.deepDischarge
+        return ret
+
+    def getRatedSOC(self):
+        """Retrieve the current state of charge as a discrete battery level predicate.
+
+        :return: The next-lower battery level corresponding to the current SOC.
+        :rtype: battery.Level
+        """
+        soc = self.getStateOfCharge()
+        lvl = self.rateSOC(soc)
+        return lvl
+
+    def getRatedSOCStr(self):
+        """Retrieve the remaining capacity as a battery level string.
+
+        :return: The next-lower battery level corresponding to the current SOC.
+        :rtype: String
+        """
+        lvl = self.getRatedSOC()
+        lvl_str = str(lvl)
+        return lvl_str
+
+    def alarmCallback(self, pinIndex):
+        # TODO: This should not be needed anymore, right?
+        # see alarmCallback from original implementation and max77960
+        pass
+
+    # Local functions for internal use
+
+    @staticmethod
+    def _transferSOC(data):
+        # LSB is 1/512 %, so shift by 9 bits.
+        ret = data + 0x0100 >> 9
+        return Percentage(ret)
+
+    @staticmethod
+    def _transferChangeRate(data):  # STC3117 exclusive
+        # LSB is 8.789 mC, scaling factor is 8.789 = 8789/1000
+        ret = (data * 8789 + 500) / 1000
+        return ret
+
+    @staticmethod
+    def _transferVoltage(data):
+        # LSB is 2.2mV, so scale by factor 2.20 = 22/10
+        ret = (data * 22 + 5) / 10
+        return Voltage(ret)
+
+    def _transferCurrent(self, data):
+        # Actually, we read out the voltage drop over the sense resistor.
+        # LSB is 5.88V, so first scaling factor is 5.88 = 294/50
+        # Value is signed!
+        # R = U/I  so we get I = U/R; Note that R is given in milliOhm!
+        # So, finally we scale by 294 / 50 * 1000 / rs = 294 * 20 / rs = 5880 / rs.
+        rs = self.REGISTER.CONFIG_GASGAUGE_0_RSENSE
+        if data >= 0:
+            ret = (data * 5880 + (rs / 2)) / rs
+        else:
+            ret = (data * 5880 - (rs / 2)) / rs
+        return ret
+
+    # /**
+    #  * AVG_CURRENT transfer function
+    #  */
+    def _transferCurrentAvg(self, data):  # STC3117 exclusive
+        if self.REGISTER.CHIP_TYPE is not ChipType.STC3117:
+            ret = Current.invalid  # Function is not implemented for this chip
+        else:
+            # Again, we actually read out the voltage drop over the sense resistor.
+            # LSB is 1.47V, partial scaling factor is 1.47 = 147/100
+            # Value is signed!
+            # Total scaling factor is: 147 / 100 * 1000 / rs = 147 * 10 / rs = 1470 / rs.
+            rs = self.REGISTER.CONFIG_GASGAUGE_0_RSENSE
+            if data >= 0:
+                ret = (data * 1470 + rs/2) / rs
+            else:
+                ret = (data * 1470 - rs/2) / rs
+            ret = Current(ret)
+        return ret
+
+    @staticmethod
+    def _transferTemperature(data):
+        # LSB is 1C, so we don't need any scaling.
```

Comment:
This must read "one degree Celsius". Not to be confused with 1C which is a charge rate.

.